### PR TITLE
new post: This Week in Glean: Looking back at Glean in 2021

### PR DESCRIPTION
**Pull request by badboy**
[Rendered](https://github.com/badboy/fnordig.de/pull/23/files?short_path=5a5bef8#diff-5a5bef82fe997f9b21b5db871e7c1b14d2fb8177370bcc68fe42dcf5ecad69c9)

### CODE REVIEW

File: _posts/2021-12-17-glean-in-2021.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+We implemented some of the new testing APIs, namely [metric coverage support][coverage-support] and [`Ping.testBeforeNextSubmit`][beforeNextSubmit].
+Glean and FOG already use these new testing capabilities, but to my knowledge other Glean users are not yet relying on that.
+Certainly an area we should make more progress on in the coming year.
+
+[coverage-support]: https://github.com/mozilla/glean/pull/1482
+[beforeNextSubmit]: https://github.com/mozilla/glean/pull/1507
+
+### UniFFI - generate all the code
+
+Migration to rely on [UniFFI] for code generation has started!
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md:75]: Maybe add one or two sentences on what UniFFI is and what it gets us?

File: _posts/2021-12-17-glean-in-2021.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+[geckoview]: https://fnordig.de/2021/09/17/glean-geckoview/
+[crash]: https://fnordig.de/2021/11/01/crashes-and-a-buggy-glean/
+
+### Firefox on Glean (FOG)
+
+We invested a lot of time in making the Firefox integration better,
+fixing bugs and convincing people to migrate their instrumentation to the new Glean-based system.
+It's going slow, but we now have the [first few important metrics][fog-dictionary] in Glean pings coming from Firefox Desktop.
+Additionally Glean is the only telemetry solution in the Desktop Background Updater.
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: I think you meant to link this?

```suggestion
Additionally Glean was used to provide telemetry for the new [Desktop Background Updater][fog-bgupdate].
```

File: _posts/2021-12-17-glean-in-2021.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+The [Glean Dictionary][dictionary] saw near-weekly releases
+and went from "promising prototype" into "valuable production tool" this year,
+thanks to a lot of work from [wlach] and [linh].
+It was the basis of several initiatives such as [metric annotations][annotations] and [tagging support][tags].
+These things will be helpful to make data at Mozilla more discoverable and better documented.
+
+The primary tool for analyzing data at Mozilla is now [Looker].
+Folks have been very busy making more and more data accessible to both data experts and non-experts alike[^1].
+Glean data is (more or less) automatically available in Looker explores.
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: You might want to link to some of the youtube videos showing how to explore Glean data in Looker: https://docs.telemetry.mozilla.org/cookbooks/looker/event_counts_explore.html

File: _posts/2021-12-17-glean-in-2021.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+Instead of bring up up adhoc dashboards for initiatives when needed we can recommend to build it all on Looker.
```
Comment (travis79) [inline _posts/2021-12-17-glean-in-2021.md]:
```suggestion
Instead of bringing up ad-hoc dashboards for initiatives when needed, we can recommend to build it all on Looker.
```

**Review by travis79 — APPROVED** [review]
One tiny nit, but this is a great post and a lot of accomplishments for our small team over a year!!

File: _posts/2021-12-17-glean-in-2021.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+[gleanjs-initial]: https://github.com/mozilla/glean.js/commit/46f028fb4ea7b8f312daf4666904c81d0a3eb171
+
+### Tooling: GLAM, Dictionary, Looker
+
+These are the tools I'm much less directly involved in,
+but they make up an important part of the Glean ecosystem.
+Afterall these are where users actually get to analyse at the data they collect.
+
+GLAM currently only serves data for Firefox Desktop using legacy telemetry data as well as Firefox for Android using Glean data.
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: Should we link to GLAM? Unfortunately GLAM is not open source but the docs page might make a good substitute: https://docs.telemetry.mozilla.org/cookbooks/glam.html

```suggestion
GLAM currently only serves data for Firefox Desktop using legacy telemetry data as well as Firefox for Android using Glean data.
```

File: _posts/2021-12-17-glean-in-2021.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+I'm excited to shift my own use away from [fiddling around with SQL][sql] to more approachable explores & dashboards,
+even for one-time analysis.
+
+[build-date]: https://bugzilla.mozilla.org/show_bug.cgi?id=1742448
+[wlach]: https://github.com/wlach
+[linh]: https://github.com/Iinh
+[annotations]: https://github.com/mozilla/glean-annotations/
+[tags]: https://bugzilla.mozilla.org/show_bug.cgi?id=1734011
+[looker]: https://looker.com/
+[sql]: https://sql.telemetry.mozilla.org/
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: Since STMO isn't public, I wonder if a link to this cookbook would be better

```suggestion
[sql]: https://docs.telemetry.mozilla.org/tools/stmo.html
```

File: _posts/2021-12-17-glean-in-2021.md (hunk 4 above)
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: Hmm, I'd suggest rewording this slightly as it reads a little strangely. Here's my attempt:

```suggestion
Instead of bringing up adhoc dashboards for initiatives when needed we can usually recommend that people just build things in Looker.
```

File: _posts/2021-12-17-glean-in-2021.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+### Glean
+
+We [released][release-page] 9 major versions of Glean and did 33 releases in total.
+We're very liberal with breaking changes & major releases,
+though we make sure that upgrades require minimal changes where possible.
+The Kotlin, Swift, Python & Rust SDKs are versioned all the same,
+so a breaking change in only one SDK will still cause a major release for other SDKs, even without breaking changes.
+
+Some more products and projects are now using Glean, including [Rally studies][rally] and the [Mozilla VPN client][vpn].
```
Comment (wlach) [inline _posts/2021-12-17-glean-in-2021.md]: Feels worth mentioning Focus here.

```suggestion
More products and projects are now using Glean, including [Rally studies][rally], the [Mozilla VPN client][vpn], and the new version of [Focus for Android](https://dictionary.telemetry.mozilla.org/apps/focus_android).
```

**Review by wlach — COMMENTED** [review]
This is great! I have some small suggestions

File: _posts/2021-12-17-glean-in-2021.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+[coverage-support]: https://github.com/mozilla/glean/pull/1482
+[beforeNextSubmit]: https://github.com/mozilla/glean/pull/1507
+
+### UniFFI - generate all the code
+
+Migration to rely on [UniFFI] for code generation has started!
+The [`uniffi` branch][uniffi-branch] is used for slowly migrating Glean piece by piece.
+Most of the metric types are [defined in UDL now][glean.udl] and the Kotlin code base has been migrated to it already,
+with all tests passing again.
+Work on this will continue next year, migrating the Swift and Python SDKs and then eventually merging it back into the main branch,
```
Comment (chutten) [inline _posts/2021-12-17-glean-in-2021.md]: ...are we going to migrate the Rust SDK as well?

File: _posts/2021-12-17-glean-in-2021.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+A full list of products is available in the [Glean Dictionary][dictionary].
+
+[release-page]: https://github.com/mozilla/glean/releases
+[rally]: https://rally.mozilla.org/
+[vpn]: https://www.mozilla.org/products/vpn/
+[dictionary]: https://dictionary.telemetry.mozilla.org/
+
+To ease integration between telemetry collected in Gecko and telemetry collected in the Android parts of the browser
+we're now [shipping Glean through GeckoView][geckoview], [we crashed it][crash], fixed it and shipped it again without new bugs cropping up.
+There's some work to be done next year to make local development easier again and to automate some of the tasks of upgading Glean across applications.
```
Comment (chutten) [inline _posts/2021-12-17-glean-in-2021.md]: The comma splices bug me a bit, and I like parentheses, but I'm not sure this is necessary.

```suggestion
To ease integration between telemetry collected in Gecko and telemetry collected in the Android parts of the browser
we're now [shipping Glean through GeckoView][geckoview]. (After [crashing it][crash], fixing it and shipping it again without new bugs cropping up.)
There's some work to be done next year to make local development easier again and to automate some of the tasks of upgading Glean across applications.
```

**Review by chutten — APPROVED** [review]
r+, optional nits

File: _posts/2021-12-17-glean-in-2021.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+> The Glean SDK is a fully self-servable telemetry SDK, usable across different platforms.
+> It enables product owners and engineers to instrument their products and rely on their data collection,
+> while following Mozilla policies & privacy standards.
+
+I'd say we're in pretty good shape here.
+The docs on [how to add Glean to a project][adding-glean] need some updates,
+but we also added a post for a slightly different audience: [Integrating Glean for product managers][glean-for-pms],
+so it's not only for engineers anymore.
+We asssisted others when they integrated Glean, but more for general data and instrumentation knowledge.
```
Comment (brizental) [inline _posts/2021-12-17-glean-in-2021.md]:
```suggestion
We assisted others when they integrated Glean, but more for general data and instrumentation knowledge.
```

File: _posts/2021-12-17-glean-in-2021.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+[adding-glean]: https://mozilla.github.io/glean/book/user/adding-glean-to-your-project/index.html
+[glean-for-pms]: https://mozilla.github.io/glean/book/user/integrating-glean-for-product-managers.html
+
+## The ideas
+
+### Metric types
+
+We specified & implemented 3 new metric types this year.
+The [`rate` metric][rate] has been implemented in the Rust SDK only and is available on Firefox Desktop.
```
Comment (brizental) [inline _posts/2021-12-17-glean-in-2021.md]:
```suggestion
The [`rate` metric][rate] has been implemented in the Rust and JavaScript SDKs and is also available on Firefox Desktop.
```

File: _posts/2021-12-17-glean-in-2021.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+The [`url` metric][url] is available in all SDKs, but not yet exposed on Firefox Desktop.
+The [`text` metric][text] is currently JavaScript-only.
+
+However the `url` and `text` metric were not usable until recently because of bugs in our schema generation tool.
+We had to apply some sort-of hacky fixes to get them to work, but the pipeline now correctly ingests data for these metrics
```
Comment (brizental) [inline _posts/2021-12-17-glean-in-2021.md]: The way this is put, it does not make us look too trustworthy, specifically "We had to apply some sort-of hacky fixes to get them to work". Not really clear to me on how to rephrase, just a thought.

**Review by brizental — APPROVED** [review]

**Review by Dexterp37 — APPROVED** [review]
